implement of api and cache

fyniny opened the pull request:
<!--  Thanks for sending a pull request! See below for tips! -->

**What this PR does / why we need it**:

this PR implement product api and cache server

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #
<!-- 填在 Fixes，PR 合并就会关 issue。填在 Reference to 会关联 issue，不会联动关闭，caicloud/quality 请用这个。-->

**Special notes for your reviewer**:

/cc @lichuan0620 

<!-- Please answer the following questions during the code freeze, and delete this line.
**Code freeze questions**

1. What causes this PR to not be merged before code freeze?
2. Why this PR is absolutely necessary for this version? Paste a screenshot of smoke testing docs if you could.
3. What's the effects after merging it?
4. Is there anyway we can skip this to not affect the overall process?
-->

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->

```release-note
NONE
```

<!--  Thanks for sending a pull request! Here are some tips:

1. https://github.com/caicloud/engineering/blob/master/docs/review_conventions.md  <-- what is the review process looks like
2. https://github.com/caicloud/engineering/blob/master/docs/commit_conventions.md  <-- how to structure your git commit
3. https://github.com/caicloud/engineering/blob/master/docs/caicloud_bot.md        <-- how to work with caicloud bot

Other tips from Kubernetes community:

1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

fyniny committed:
feat(*): add cache server

fyniny committed:
feat(*): add cache server

fyniny committed:
chore(gitignore): add bin vendor to gitignore

fyniny committed:
Merge pull request #1 from fyniny/feat-cache-server

chore(gitignore): add bin vendor to gitignore

fyniny committed:
feat(cache client): compelete cache client

fyniny committed:
feat(cache client): compelete cache client

fyniny committed:
style(*): add annotation and change user email

fyniny committed:
style(*): add annotation and change user email

fyniny committed:
Merge branch 'master' of github.com:fyniny/nirvana-practice

fyniny committed:
feat(Dockerfile): contianerize cache & server

fyniny committed:
feat(Dockerfile): contianerize cache & server

fyniny committed:
refactor(*): refactor Dockerfile of server & cache

fyniny committed:
refactor(*): refactor Dockerfile of server & cache

bbbmj committed:
chore(docs): update PULL_REQUEST_TEMPLATE.md

lichuan0620 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -1,16 +1,16 @@
-<!-- Issue template; please answer the questions. -->
```
commit 的时候读一遍自己在 commit 的代码；这几个文件你肯定是不想改的（看到也没改，应该是 IDE 做得格式改动）

bbbmj committed:
chore(docs): update PULL_REQUEST_TEMPLATE.md

hezhizhen committed:
fix(*): add gitignore (#3)

lichuan0620 commented on pkg/apis/cache/descriptors/cache.go:
```diff
@@ -0,0 +1,119 @@
+				Parameters: []definition.Parameter{
+					{
+						Source:      definition.Query,
+						Name:        "limit",
```
limit 这种一般会有一个 list option，里面包含 start limit 这些通用参数。一个一个定义代码很不方便看，写起来也费力

fyniny replied:
好的，我修正了。CacheClient改为接口类型。传神有空在帮我看看哪里需要再优化的

lichuan0620 commented on pkg/apis/cache/descriptors/cache.go:
```diff
@@ -0,0 +1,119 @@
+							definition.Parameter{
+								Source:      definition.Path,
+								Name:        "product",
+								Operators:   []definition.Operator{
```
用 operator 一般是为了执行校验逻辑，一个 string 就不需要了，直接 PathParameterFor 就好了

lichuan0620 commented on pkg/apis/install.go:
```diff
@@ -1,23 +1,56 @@
-
-	"github.com/caicloud/nirvana-practice/pkg/apis/v1alpha1/descriptors"
+	"github.com/caicloud/nirvana/rest"
+	"sync"
```
import 注意格式。原生 package 在先，然后是外部 package，然后是内部的

fyniny committed:
style(*): rebase from caicloud/master

fyniny committed:
refactor(cacheClient): change type CacheClient from struct to interface

change CacheClient into interface, new CacheClient is name cacheClient. Make some code specification

lichuan0620 commented:
/close
